File: app/src/main/java/com/lelloman/demo/fragment/ClassicTilesFragment.java
package com.lelloman.demo.fragment;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.lelloman.identicon.drawable.ClassicIdenticonTile;
import com.lelloman.identicon.util.TileMeasures;

@SuppressWarnings("ResourceType")
public class ClassicTilesFragment extends Fragment {

	public static final int TILE_VIEW_ID = 1;
	public static final int TEXT_VIEW_ID = 2;

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

		RecyclerView recyclerView = new RecyclerView(getContext());

		recyclerView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
		recyclerView.setAdapter(new TilesAdapter());
		recyclerView.setLayoutManager(new LinearLayoutManager(getContext(),LinearLayoutManager.VERTICAL,false));


		return recyclerView;
	}

	public class TilesAdapter extends RecyclerView.Adapter<ViewHolder> {

		@Override
		public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

			LinearLayout view = new LinearLayout(getContext());
			view.setOrientation(LinearLayout.HORIZONTAL);
			view.setBackgroundColor(0xff999999);
			view.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

			TileView tileView = new TileView(getContext());
			int tileSize = (int) (getContext().getResources().getDisplayMetrics().density * 32);
			int tileMargin = tileSize / 4;
			LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(tileSize, tileSize);
			layoutParams.setMargins(tileMargin, tileMargin, tileMargin, tileMargin);
			tileView.setLayoutParams(layoutParams);
			tileView.setId(TILE_VIEW_ID);

			view.addView(tileView);

			TextView textView = new TextView(getContext());
			textView.setId(TEXT_VIEW_ID);

			layoutParams = new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT);
			layoutParams.weight = 1;
			layoutParams.gravity = Gravity.CENTER;
			textView.setLayoutParams(layoutParams);

			view.addView(textView);

			return new ViewHolder(view);
		}

		@Override
		public void onBindViewHolder(ViewHolder holder, int position) {

			ClassicIdenticonTile.Tiles tile = ClassicIdenticonTile.INSTANCE.getAll()[position];
			holder.tileView.setTile(tile);

			String text = String.format("%s - %s%s", position, tile.name(), position % 4 == 0 ? "*" : "");
			holder.textView.setText(text);

		}

		@Override
		public int getItemCount() {
			return ClassicIdenticonTile.INSTANCE.getAll().length;
		}
	}

	public static class ViewHolder extends RecyclerView.ViewHolder {
		public final TileView tileView;
		public final TextView textView;

		public ViewHolder(View view){
			super(view);
			tileView = (TileView) view.findViewById(TILE_VIEW_ID);
			textView = (TextView) view.findViewById(TEXT_VIEW_ID);
		}
	}

	public static class TileView extends View {

		Paint whitePaint, blackPaint;
		ClassicIdenticonTile.Tiles tile;

		public TileView(Context context) {
			super(context);

			whitePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
			whitePaint.setColor(0xffffffff);
			whitePaint.setStyle(Paint.Style.FILL);

			blackPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
			blackPaint.setColor(0xff000000);
			blackPaint.setStyle(Paint.Style.FILL);
		}

		public void setTile(ClassicIdenticonTile.Tiles tile){
			this.tile = tile;
			invalidate();
		}

		@Override
		protected void onDraw(Canvas canvas) {
			super.onDraw(canvas);
			if(tile == null)
				return;

			TileMeasures measures = new TileMeasures(canvas.getWidth(),canvas.getHeight());

			tile.draw(canvas,measures,0,whitePaint,blackPaint);
		}
	}
}

File: app/src/main/java/com/lelloman/demo/fragment/ClassicTilesFragment.kt
package com.lelloman.demo.fragment

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import com.lelloman.identicon.drawable.ClassicIdenticonTile
import com.lelloman.identicon.util.TileMeasures

class ClassicTilesFragment : Fragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val recyclerView = RecyclerView(requireContext())

        recyclerView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        recyclerView.adapter = TilesAdapter()
        recyclerView.layoutManager = GridLayoutManager(context, 4)


        return recyclerView
    }

    inner class TilesAdapter : RecyclerView.Adapter<ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

            val view = LinearLayout(context)
            view.orientation = LinearLayout.VERTICAL
            view.setBackgroundColor(-0x666667)
            view.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            view.gravity = Gravity.CENTER

            val tileView = TileView(requireContext())
            val tileSize = (requireContext().resources.displayMetrics.density * 48).toInt()
            val tileMargin = tileSize / 6
            var layoutParams = LinearLayout.LayoutParams(tileSize, tileSize)
            layoutParams.setMargins(tileMargin, tileMargin, tileMargin, 0)
            tileView.layoutParams = layoutParams
            tileView.id = TILE_VIEW_ID

            view.addView(tileView)

            val textView = TextView(context)
            textView.id = TEXT_VIEW_ID
            textView.textSize = 10f
            textView.gravity = Gravity.CENTER

            layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            layoutParams.setMargins(0, 0, 0, tileMargin)
            textView.layoutParams = layoutParams

            view.addView(textView)

            return ViewHolder(view)
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            ClassicIdenticonTile.all[position].let { tile ->
                holder.tileView.setTile(tile)

                val suffix = if (position % 4 == 0) "*" else ""
                holder.textView.text = "$position$suffix"
            }
        }

        override fun getItemCount(): Int {
            return ClassicIdenticonTile.all.size
        }
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val tileView: TileView = view.findViewById<View>(TILE_VIEW_ID) as TileView
        val textView: TextView = view.findViewById<View>(TEXT_VIEW_ID) as TextView
    }

    class TileView(context: Context) : View(context) {

        private var whitePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = -0x1
            style = Paint.Style.FILL
        }
        private var blackPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = -0x1000000
            style = Paint.Style.FILL
        }
        private var tile: ClassicIdenticonTile.Tiles? = null

        fun setTile(tile: ClassicIdenticonTile.Tiles) {
            this.tile = tile
            invalidate()
        }

        @SuppressLint("CanvasSize", "DrawAllocation")
        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            tile?.let { tile ->
                val measures = TileMeasures(canvas.width, canvas.height)
                tile.draw(canvas, measures, 0, whitePaint, blackPaint)
            }
        }
    }

    companion object {
        const val TILE_VIEW_ID = 1
        const val TEXT_VIEW_ID = 2
    }
}
